### ISSUE TITLE
Support for different dimension vector for DCCA

### ISSUE THREAD

**Issue by AdirRahamim** [jameschapman19/cca_zoo#72]
Hi, 

I'm the current version, DCCA only works if the two vectors have the same dimensionality, however this is not a necessary assumption as DCCA also works for different size vectors. Is there a possibility to add support for different dimension vector in DCCA?

**Comment by jameschapman19** [comment]
Hi @AdirRahamim.

This should work fine. Will look something like:

    encoder_1 = architectures.Encoder(latent_dims=latent_dims, feature_size=FEATURESIZE1)
    encoder_2 = architectures.Encoder(latent_dims=latent_dims, feature_size=FEATURESIZE2)

    dcca_model = DCCA(
        latent_dims=latent_dims,
        encoders=[encoder_1, encoder_2],
        objective=objectives.CCA,
    )

    dcca_model = DeepWrapper(dcca_model, device=device)
    dcca_model.fit(train_dataset, epochs=3)

**Comment by AdirRahamim** [comment]
Hi,
Yes, I tried doing so, however, in the data loader you try to stack the two vectors and return them as data where the first vector is data[0] and the second is data[1]:

`self.model.train()
        train_loss = 0
        for batch_idx, (data, label) in enumerate(train_dataloader): # <- Here error occures
            data = [d.double().to(self.device) for d in list(data)]
            loss = self._update_weights(*data)
            train_loss += loss.item()
        return train_loss / len(train_dataloader)`

This is the code of _train_epoch under DeepWrapper. In the case where they are not of the same dimensions, you can't stack them and the data fetch arises an error. I think you might consider a change in the dataloader so instead of stacking the vectors, just return data0 and data1.
Thank you for your help!

**Comment by jameschapman19** [comment]
Ah ok I've assumed that data is a tuple as in the below example. That being said I currently don't test the views with different dimensions case so will use this as a nudge to do that. For me this works fine (e.g. changing the dimensionality of the data for the test cases and using my basic CCA_Dataset)

class CCA_Dataset(Dataset):
    """
    Class that turns numpy arrays into a torch dataset

    """

    def __init__(self, views, labels=None):
        """

        :param views: list/tuple of numpy arrays or array likes with the same number of rows (samples)
        :param labels: optional labels
        """
        self.views = [view for view in views]
        if labels is None:
            self.labels = np.zeros(len(self.views[0]))
        else:
            self.labels = labels

    def __len__(self):
        return len(self.labels)

    def __getitem__(self, idx):
        label = self.labels[idx]
        views = [view[idx].astype(np.float32) for view in self.views]
        return tuple(views), label

**Comment by jameschapman19** [comment]
Another route which I mentioned in previous issue is checking out the CCALightning class in the pytorchlightning branch on this repo.

The original idea of the DeepWrapper was to be able to build an object which could then have transform/score methods much like scikit-learn models and the other CCA models here. You could in principle inherit the class and overwrite the _train_epoch method which I think was my original vision. What I've since realised is that pytorch-lightning does the training loop abstraction much better than I could hope to so next version will build off their framework and make it super easy to do custom cases.

**Comment by AdirRahamim** [comment]
OK thank you for your help! I will try to rewrite the relevant parts of the code so it will support the function.

**Comment by jameschapman19** [comment]
It already works for the case you are describing in the way I have described

**Comment by jameschapman19** [comment]
Something must not be clear about my description so I can learn something from your problem 😊

**Comment by AdirRahamim** [comment]
Maybe there's something I didn't understand from your answer. Currently, this is test code I'm running, and falls with the error I previously mentioned:

```
a = np.random.randn(2000, 50)
b = np.random.randn(2000, 100)
m1 = min(a.shape[1], b.shape[1])
train_dataset = data.CCA_Dataset([a, b])
encoder_a = Encoder(latent_dims=m1, feature_size=50, layer_sizes=[128, 256])
encoder_b = Encoder(latent_dims=m1, feature_size=100, layer_sizes=[128, 256])
dcca = DCCA(latent_dims=m1, objective=objectives.CCA, encoders=[encoder_a, encoder_b])
dcca = DeepWrapper(dcca, device='cpu').fit(train_dataset) 
```

What should I change in order to make it run properly?

**Comment by jameschapman19** [comment]
another weird one where your code runs without error for me! What's the error at that line?

**Comment by AdirRahamim** [comment]
That's really strange! The error is at the last line, in the fit function.
As I wrote in the third post, I get an error when it tries to fetch the data and stack the two vectors, however, it arises an error that it cannot stack vectors of different dimensions.

**Comment by jameschapman19** [comment]
clone of this repo or published from pip?

**Comment by AdirRahamim** [comment]
From pip, now tried to clone and run, and it works, I guess there's a difference between the published pip version and git version

**Comment by jameschapman19** [comment]
ah ok got it. Well that's good in a way and bad in another way! I have published a new version (v.1.9.1) but I can't see anything which addresses the line in question. I would be curious to know if this version installed from pip fixes your problem. If so I can close this.

**Comment by AdirRahamim** [comment]
Now it raises the following error:
"Reshape your data either using array.reshape(-1, 1) if your data has a single feature or array.reshape(1, -1) if it contains a single sample."
Even when calling the regular cca method

**Comment by jameschapman19** [comment]
Passes all the tests so seems as though it's your end but it would be good to understand. I assume in a new environment you don't see the error (e.g. running the script in a Google colab)?

**Comment by AdirRahamim** [comment]
I think t's something with package versions. I will try to fix it.
Thank you for your help!

**Comment by AdirRahamim** [comment]
OK, I tried on fresh new environment, I still get this error while running the following simple code that worked before:

```
a = np.random.randn(2000, 50)
b = np.random.randn(2000, 100)
m1 = min(a.shape[1], b.shape[1])
basic_cca = CCA(latent_dims=m1)
basic_cca.fit(a, b) 
```

it seems like something went wrong after your last commit.

**Comment by jameschapman19** [comment]
.fit((a,b))

**Comment by AdirRahamim** [comment]
OK, now everything is working perfectly! Thank you for being patient.
